cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ETLStage.java

```diff
@@ -77,6 +104,117 @@ public ETLStage upgradeStage(UpgradeContext upgradeContext) {
+                                                        outputSchema);
+        default:
+          return this;
+        }
+      }
+
+    // No update action provided so return stage as is.
+    return this;
+  }
+
+  /**
+   * Upgrade plugin used in the stage.
+   * 1. If plugin is using fixed version and a new plugin artifact is found with higher version in SYSTEM scope,
+   *    use the new plugin.
+   * 2. If plugin is using a plugin range and a new plugin artifact is found with higher version in SYSTEM scope,
+   *    move the upper bound of the range to include the new plugin artifact. Also change plugin scope.
+   *    If new plugin is in range, do not change range. (Note: It would not change range even though new plugin is in
+   *    different scope).
+   *
+   * @param updateContext To use helper functions like getPluginArtifacts.
+   * @return Updated plugin object to be used for the udated stage. Returned null if no changes to current plugin.
+   */
+  private ETLPlugin upgradePlugin(ApplicationUpdateContext updateContext) throws Exception {
+    // Currently tries to find latest plugin in SYSTEM scope and upgrades current plugin if version is higher,
+    // ignoring current plugin scope.
+    // In future, we can modify logic to fetch the latest plugin in any scope.
+    List<ArtifactId> candidates =
+      updateContext.getPluginArtifacts(plugin.getType(), plugin.getName(),
+                                       ArtifactScope.SYSTEM, null);
+    if (candidates.isEmpty()) {
+      return plugin;
+    }
+
+    // getPluginArtifacts returns plugins sorted in ascending order.
+    // TODO: Consider passing sort order as parameter.
+    ArtifactId newPlugin = candidates.get(candidates.size() - 1);
+    String newVersion = getUpgradedVersionString(newPlugin);
+    // If getUpgradedVersionString returns null, candidate plugin is not valid for upgrade.
+    if (newVersion == null) {
+      return plugin;
+    }
+
+    ArtifactSelectorConfig newArtifactSelectorConfig =
+      new ArtifactSelectorConfig(newPlugin.getScope().name(), newPlugin.getName(),
+                                 newVersion);
+    io.cdap.cdap.etl.proto.v2.ETLPlugin upgradedEtlPlugin =
+      new io.cdap.cdap.etl.proto.v2.ETLPlugin(plugin.getName(), plugin.getType(),
+                                              plugin.getProperties(),
+                                              newArtifactSelectorConfig);
+    return upgradedEtlPlugin;
+  }
+
+  /**
+   * Returns new valid version string for plugin upgrade if any changes are required. Returns null if no change to
+   * current plugin version.
+   * Artifact selector config only stores plugin version as string, it can be either fixed version or range.
+   * Hence, if the plugin version is fixed, replace the fixed version with newer fixed version. If it is a range,
+   * move the upper bound of the range to the newest version.
```

Comment:
if they've explicitly given a range with an upper bound, is it really correct to increase that range? 

To me, it makes more sense to choose the highest artifact within that range.

Reply:
it would be useful to give a couple inputs and outputs

Reply:
If we dont change the range (or plan to change the range), than in that case we would never touch ranged version here right? Than we leave it to deploy to pick the highest version of plugin available in the range. Then even if a new version of plugin (outside current range) is introduced it wont get picked up.

IMO, we should increase the range so new plugin gets picked up. There are boundary cases with scoping that we need to figure out, but unless we upgrade the range, not sure how a new plugin can get picked up. Feel free to correct me if I am missing something.